### ISSUE TITLE
BG2: dual-class has wrong available proficiencies 

### ISSUE THREAD

**Issue by Tomsod** [gemrb/gemrb#1732]
So, in the original game, an activated dual-class combines available weapon proficiencies from both classes in a complicated way.  In GemRB, only the new class counts.  This is wrong, but I don't know what precisely the correct algorithm is.  Sometimes it's permissive (Fighter/Thief got all weapons), sometimes restrictive (Fighter/Cleric only has blunt weapons), sometimes even both (the infamous Kensai/Mage has all melee weapons, but not slings or darts!)  A relatively simple approximation would be to use the first class in order priest>warrior>rogue>mage, but it still fails for some kits (e.g. beastmaster/cleric).  This really needs research.

**Comment by Tomsod** [comment]
Or maybe allow all weapons available to either class except those forbidden by any present warrior or priest class/kit.  Rogue and mage kits conveniently never have weapon disadvantages.

**Comment by lynxlynxlynx** [comment]
So you mean at level up, right?
BG1 and IWD have https://gibberlings3.github.io/iesdp/files/2da/2da_tots/clasweap.htm
BG2 has https://gibberlings3.github.io/iesdp/files/2da/2da_tob/weapprof.htm

We handle both already, so if something is wrong and it's not wrong in the data, it's a bug in SetupProfsWindow.

#1072

**Comment by Tomsod** [comment]
Oh, yes, it is SetupProfsWindow.  Should have clarified.  It just doesn't account for the old class at all.  I've "fixed" this for myself by combining the available weapons, which works for Imoen, but then I realized it's much more complicated, thus no PR.

**Comment by Tomsod** [comment]
And before you ask: no, you cannot use multiclass table entries either, because e.g. fighter duals can advance to 5 stars while fighter multis are limited to 2.  Forgot to write this before.

**Comment by burner1024** [comment]
Actually, original BG2 does allow kensai/mage to put pips into sling and dart. But then it doesn't allow to use them :/.
It doesn't allow put pips into bows, which both classes are unable to use.

If I understand the algo correctly, it's:
1) Combine proficiencies from both base classes. Clerics and druids are restricted by ethos and thus can only get their options reduced, not expanded. The rest of the combos are combined permissively.
2) Apply kit specific restrictions on top. (Can't remember if there are kit specific permissions, hopefully not).

Or, going another way, use multiclass tables for availability, base class tables to determine max allowed (max of both classes), and then further restrict by kit.

Although, letting duals advance over the 2nd class limit seems like a bug, or at very least, oversight in the first place.

**Comment by Tomsod** [comment]
Hmm, if the original game doesn't catch the Kensage case after all, then the algorithm could've been as simple as "priests restrictive, everyone else permissive".  Of course, we can do better.  I've just re-checked, only warrior kits have special weapon restrictions, while Swashbucklers and Blades can actually advance two-weapon style farther than their base classes, so I still propose the "priests and warrior kits restrictive, rogues and mages and their kits permissive" approach.  (The fact that base warrior classes can learn all weapons and styles also helps.)

Old class weapon skills being available to duals makes sense if you think of them as an innate feature of the old class (which is thus regained when the class is reactivated) rather than something gained by leveling.  At least, the proficiency rate is taken from the new class., which is fair enough.

**Comment by burner1024** [comment]
Yes, swashbucklers can actually advance farther than base. But base can still pup 1 pip into 2w style - it's the max that changes, not availability. Bards can't dual, so in original game it doesn't matter.

It does bring up another point: would probably want still want to display something sensible, or at least not break for characters with illegal dual/multi class (bugs, mods, shadowkeeper, etc). So multiclass tables aren't reliable to determine availability.
Is monk a priest or a warrior or a rogue in this scenario?

What simplifies it in the original is that all weapon styles are available for both "restrictive" classes, clerics and druids, and you can't dual into a kit, only from.
While in the linked issue there's discussion about allowing to dual into a kit, and even double kit. Now, think berserker/archer or monk/swashbuckler... On the other hand, who goes down this route has only themselves to blame =).

In the end "priests and warrior kits restrictive, rogues and mages and their kits permissive" sounds about right, with a special clause that priests are only restrictive on availability, while warrior kits are restrictive on the maximum too.
